File: chroma-league/src/main/java/net/booone/chromaleague/hud/animations/LevelUpAnimation.java
package net.booone.chromaleague.hud.animations;

import net.booone.chromaleague.state.GameStateHelper;
import net.booone.razersdk.animation.AnimatedFrame;
import net.booone.razersdk.animation.Animation;
import net.booone.razersdk.animation.Frame;
import net.booone.razersdk.animation.SimpleFrame;
import net.booone.razersdk.color.StaticColor;
import net.booone.razersdk.sdk.ProgressiveRzKeySelector;

import static net.booone.razersdk.sdk.RzKey.*;

public class LevelUpAnimation extends Animation {
    private int previousLevel;

    @Override
    public Frame getFrame() {
        addLevelUpAnimationIfNeeded();
        if (hasFrame()) {
            return super.getFrame();
        }
        return new SimpleFrame().getFrame();
    }

    private void addLevelUpAnimationIfNeeded() {
        if (GameStateHelper.getLevel() > previousLevel) {
            levelUp();
            previousLevel = GameStateHelper.getLevel();
        }
    }

    void levelUp() {
        ProgressiveRzKeySelector levelUpSelector = new ProgressiveRzKeySelector.Builder()
                .addPart(RZKEY_MACRO5, RZKEY_LCTRL, RZKEY_LWIN)
                .addPart(RZKEY_MACRO4, RZKEY_LSHIFT)
                .addPart(RZKEY_MACRO3, RZKEY_CAPSLOCK)
                .addPart(RZKEY_MACRO2, RZKEY_TAB)
                .addPart(RZKEY_MACRO1, RZKEY_TILDE)
                .addPart(RZKEY_ESC)
                .withLength(3)
                .build();
        final AnimatedFrame frame = new AnimatedFrame();
        for (int i = 0; i < levelUpSelector.getTotalSteps(); i++) {
            frame.addAnimationFrame(2, new SimpleFrame(levelUpSelector.getNextPart(), StaticColor.YELLOW));
        }
        addToFront(frame);
    }
}

File: src/hud/animations/LevelUpAnimation.ts
import {AnimatedFrame} from '../../razer-sdk/animation/AnimatedFrame.js';
import {Animation} from '../../razer-sdk/animation/Animation.js';
import {Frame} from '../../razer-sdk/animation/Frame.js';
import {SimpleFrame} from '../../razer-sdk/animation/SimpleFrame.js';
import {StaticColor} from '../../razer-sdk/color/StaticColor.js';
import {ProgressiveRzKeySelectorBuilder} from '../../razer-sdk/sdk/ProgressiveRzKeySelector.js';
import {RzKey} from '../../razer-sdk/sdk/RzKey.js';
import * as GameStateHelper from '../../state/GameStateHelper.js';

export class LevelUpAnimation extends Animation {
    private previousLevel = 0;

    pollFrame(): Frame {
        this.addLevelUpAnimationIfNeeded();
        if (this.hasFrame()) {
            return super.pollFrame();
        }
        return new SimpleFrame().pollFrame();
    }

    private addLevelUpAnimationIfNeeded(): void {
        if (GameStateHelper.getLevel() > this.previousLevel) {
            this.levelUp();
            this.previousLevel = GameStateHelper.getLevel();
        }
    }

    levelUp(): void {
        const levelUpSelector = new ProgressiveRzKeySelectorBuilder()
            .addPartFromKeys(RzKey.RZKEY_MACRO5, RzKey.RZKEY_LCTRL, RzKey.RZKEY_LWIN)
            .addPartFromKeys(RzKey.RZKEY_MACRO4, RzKey.RZKEY_LSHIFT)
            .addPartFromKeys(RzKey.RZKEY_MACRO3, RzKey.RZKEY_CAPSLOCK)
            .addPartFromKeys(RzKey.RZKEY_MACRO2, RzKey.RZKEY_TAB)
            .addPartFromKeys(RzKey.RZKEY_MACRO1, RzKey.RZKEY_TILDE)
            .addPartFromKeys(RzKey.RZKEY_ESC)
            .withLength(3)
            .build();
        const frame = new AnimatedFrame();
        for (let i = 0; i < levelUpSelector.getTotalSteps(); i++) {
            frame.addAnimationFrame(2, new SimpleFrame([...levelUpSelector.getNextPart()], StaticColor.YELLOW));
        }
        this.addToFront(frame);
    }
}
